[X86] Remove extra MOV after widening atomic load

jofrn opened the pull request:
This change adds patterns to optimize out an extra MOV
present after widening the atomic load.

---

**Stack**:
- #120716
- #142320
- #138635 ⬅
- #120598
- #120387
- #120386


⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

RKSimon commented on llvm/lib/Target/X86/X86InstrCompiler.td:
```diff
@@ -1200,6 +1200,13 @@ def : Pat<(i16 (atomic_load_nonext_16 addr:$src)), (MOV16rm addr:$src)>;
 def : Pat<(i64 (atomic_load_nonext_64 addr:$src)), (MOV64rm addr:$src)>;
 
+def : Pat<(v4i32 (scalar_to_vector (i32 (anyext (i16 (atomic_load_16 addr:$src)))))),
+           (MOVDI2PDIrm addr:$src)>;   // load atomic <2 x i8>
```
this will dereference 32-bits

jofrn replied:
Switched it to a `zext` and now it dereferences 16 bits in the asm.

RKSimon replied:
Next thing is to add SSE/AVX handling - I've added better test coverage at d27d0c7a5266f89f9d62464e71be98421aae598d

jofrn replied:
Without loss of generality, do we not need the `v2` in `--check-prefixes=CHECK,CHECKv2-O0` due to divergence of asm? https://github.com/llvm/llvm-project/blob/d27d0c7a5266f89f9d62464e71be98421aae598d/llvm/test/CodeGen/X86/atomic-load-store.ll#L7

RKSimon replied:
Yes, you'll have to add extra check-prefixes - base + v2 can share CHECK-SSE-O* and  v3/4 can share a CHECK-AVX-O* prefixes

jofrn committed:
[X86] Remove extra MOV after widening atomic load

This change adds patterns to optimize out an extra MOV
present after widening the atomic load.

commit-id:45989503
